language_service/src/completion/path_context.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc::{
+    ast::{
+        visit::{self, Visitor},
+        Attr, Block, CallableDecl, Expr, ExprKind, FieldAssign, FieldDef, FunctorExpr, Ident, Item,
+        ItemKind, Namespace, Package, Pat, Path, QubitInit, SpecDecl, Stmt, StructDecl, Ty, TyDef,
+        TyKind,
+    },
+    parse::completion::PathKind,
+};
+use std::rc::Rc;
+
+/// Provides the qualifier and the expected name kind for the
+/// incomplete path (e.g. `foo.bar.`) at the cursor offset.
+///
+/// Methods may panic if the offset does not fall within an incomplete path.
+#[derive(Debug)]
+pub(super) struct IncompletePath<'a> {
+    qualifier: Option<&'a [Ident]>,
+    context: Option<PathKind>,
+    offset: u32,
+}
+
+impl<'a> IncompletePath<'a> {
+    pub fn init(offset: u32, package: &'a Package) -> Self {
+        let mut offset_visitor = OffsetVisitor {
+            offset,
+            visitor: IncompletePath {
+                offset,
+                context: None,
+                qualifier: None,
+            },
+        };
+
+        offset_visitor.visit_package(package);
+
+        offset_visitor.visitor
+    }
+}
+
+impl<'a> Visitor<'a> for IncompletePath<'a> {
+    fn visit_item(&mut self, item: &Item) {
+        match *item.kind {
+            ItemKind::Open(..) => self.context = Some(PathKind::Namespace),
+            ItemKind::ImportOrExport(..) => self.context = Some(PathKind::Import),
+            _ => {}
+        }
+    }
+
+    fn visit_ty(&mut self, ty: &Ty) {
+        if let TyKind::Path(..) = *ty.kind {
+            self.context = Some(PathKind::Ty);
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        if let ExprKind::Path(..) = *expr.kind {
+            self.context = Some(PathKind::Expr);
+        } else if let ExprKind::Struct(..) = *expr.kind {
+            self.context = Some(PathKind::Struct);
+        }
+    }
+
+    fn visit_path_kind(&mut self, path: &'a qsc::ast::PathKind) {
+        self.qualifier = match path {
+            qsc::ast::PathKind::Ok(path) => path.segments.as_ref().map(AsRef::as_ref),
+            qsc::ast::PathKind::Err(Some(incomplete_path)) => Some(&incomplete_path.segments),
+            qsc::ast::PathKind::Err(None) => None,
+        };
```

Comment:
nit: clean up with use to match rest of file

Reply:
Sad story: there are two different structs named `PathKind` in this file, and the qualifiers here are to disambiguate. Maybe I'll use an alias.